### PR TITLE
Implement more rendering options as specialization constants

### LINKED ISSUE #2247: Add a property to change the texture sampling mode for Decals, OmniLight3D and SpotLight3D projectors
### Describe the project you are working on

The Godot editor (Vulkan) :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

There's currently no way to change the texture sampling mode used for decals and light projectors. This means all decals and light projectors use linear filtering with no anisotropic filtering. There are two downsides to this:

- The lack of anisotropic filtering will make decals look blurry when viewed at oblique angles.
- For 3D games with a pixel art style, decals will look blurry instead of being pixelated.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add property to change the texture sampling mode for Decals, OmniLight3D and SpotLight3D projectors (nearest, linear, mipmaps, anisotropic, …).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Ideally, this should be adjustable on a per-node basis. The Sampling property would have the same values as the one used in StandardMaterial3D. However, since decals and projectors are part of an atlas, I understand this may not be possible. In this case, this option would have to be exposed as a project setting and would affect all decals/projectors used in the project.

*Note that light projectors have to be reimplemented first: https://github.com/godotengine/godot/issues/45733*

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, as Godot 4.0's decal rendering is implemented by the engine and can't be overridden in a script.

### Is there a reason why this should be core and not an add-on in the asset library?

This is core renderer functionality.

### PR DESCRIPTION
**Pull request by reduz**
* Shadow quality settings now specialization constant.
* Decal and light projector quality too.
* Changing those settings forces re-creation of the pipelines.

These changes should help improve performance related to shadow mapping, and allows improving performance by sacrificing decal and light projector quality.

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/2247.*

### CODE REVIEW

**Comment by clayjohn** [comment]
Looks like there are a few issues with CI:
1. docs need to be generated and added
2. some override functions are not marked as override. 

Other than that, the implementation looks good.

**Comment by Calinou** [comment]
I benchmarked this PR with [godot-rendering-performance-test](https://github.com/Calinou/godot-rendering-performance-test) using the `--print-fps` command line argument. The test scene uses the default soft shadow settings but has no decals.

### Before

```yaml
# 1024×600
Project FPS: 509 (1.9 mspf)
Project FPS: 485 (2.0 mspf)
Project FPS: 500 (2.0 mspf)
Project FPS: 506 (1.9 mspf)
Project FPS: 489 (2.0 mspf)
Project FPS: 486 (2.0 mspf)
Project FPS: 496 (2.0 mspf)

# 2560×1440
Project FPS: 145 (6.8 mspf)
Project FPS: 145 (6.8 mspf)
Project FPS: 147 (6.8 mspf)
Project FPS: 148 (6.7 mspf)
```

### After

```yaml
# 1024×600
Project FPS: 545 (1.8 mspf)
Project FPS: 544 (1.8 mspf)
Project FPS: 556 (1.7 mspf)
Project FPS: 556 (1.7 mspf)
Project FPS: 545 (1.8 mspf)
Project FPS: 556 (1.7 mspf)

# 2560×1440
Project FPS: 162 (6.1 mspf)
Project FPS: 163 (6.1 mspf)
Project FPS: 163 (6.1 mspf)
```

File: doc/classes/ProjectSettings.xml [hunk 1]
```diff
@@ -1327,6 +1327,10 @@
 		</member>
 		<member name="rendering/2d/snap/snap_2d_vertices_to_pixel" type="bool" setter="" getter="" default="false">
 		</member>
+		<member name="rendering/anti_aliasing/decals/high_quality" type="bool" setter="" getter="" default="true">
+		</member>
+		<member name="rendering/anti_aliasing/light_projectors/high_quality" type="bool" setter="" getter="" default="true">
+		</member>
```
Comment (Calinou) [inline doc/classes/ProjectSettings.xml:1333]:
```suggestion
		<member name="rendering/anti_aliasing/decals/high_quality" type="bool" setter="" getter="" default="true">
			If [code]true[/code], decals will be sampled with [code]textureGrad()[/code], which results in a smoother, less grainy appearance for distant decals. However, this has a performance cost. If [code]false[/code], decals will be sampled with [code]textureLod()[/code] instead, which is faster but may result in distant decals looking grainy.
			[b]Note:[/b] This property is only read when the project starts. To change this value at run-time, use [method RenderingServer.decals_set_quality].
		</member>
		<member name="rendering/anti_aliasing/light_projectors/high_quality" type="bool" setter="" getter="" default="true">
			If [code]true[/code], light projector textures will be sampled with [code]textureGrad()[/code], which results in a smoother, less grainy appearance for distant light projector textures. However, this has a performance cost. If [code]false[/code], light projector textures will be sampled with [code]textureLod()[/code] instead, which is faster but may result in distant light projector textures looking grainy.
			[b]Note:[/b] This property is only read when the project starts. To change this value at run-time, use [method RenderingServer.light_projectors_set_quality].
		</member>
```

File: doc/classes/RenderingServer.xml [hunk 2]
```diff
@@ -1209,6 +1209,14 @@
 			<description>
 			</description>
 		</method>
+		<method name="decals_set_quality">
+			<return type="void">
+			</return>
+			<argument index="0" name="high_quality" type="bool">
+			</argument>
+			<description>
+			</description>
```
Comment (Calinou) [inline doc/classes/RenderingServer.xml:1218]:
```suggestion
			<description>
				Sets the decal rendering quality. If [code]high_quality[/code] is [code]true[/code], decals will be sampled with [code]textureGrad()[/code], which results in a smoother, less grainy appearance for distant decals. However, this has a performance cost. If [code]high_quality[/code] is [code]false[/code], decals will be sampled with [code]textureLod()[/code] instead, which is faster but may result in distant decals looking grainy.
				See also [member ProjectSettings.rendering/anti_aliasing/decals/high_quality].
			</description>
```

File: doc/classes/RenderingServer.xml [hunk 3]
```diff
@@ -2153,6 +2161,14 @@
 				Sets whether to use a dual paraboloid or a cubemap for the shadow map. Dual paraboloid is faster but may suffer from artifacts. Equivalent to [member OmniLight3D.omni_shadow_mode].
 			</description>
 		</method>
+		<method name="light_projectors_set_quality">
+			<return type="void">
+			</return>
+			<argument index="0" name="high_quality" type="bool">
+			</argument>
+			<description>
+			</description>
```
Comment (Calinou) [inline doc/classes/RenderingServer.xml:2170]:
```suggestion
			<description>
				Sets the light projector texture sampling quality. If [code]high_quality[/code] is [code]true[/code], light projector textures will be sampled with [code]textureGrad()[/code], which results in a smoother, less grainy appearance for distant light projector textures. However, this has a performance cost. If [code]high_quality[/code] is [code]false[/code], light projector textures will be sampled with [code]textureLod()[/code] instead, which is faster but may result in distant light projector textures looking grainy.
				See also [member ProjectSettings.rendering/anti_aliasing/decals/high_quality].
			</description>
```

**Review by Calinou — DISMISSED** [review]

File: editor/editor_node.cpp [hunk 4]
```diff
@@ -508,6 +508,9 @@ void EditorNode::_update_from_settings() {
 
 	float lod_threshold = GLOBAL_GET("rendering/mesh_lod/lod_change/threshold_pixels");
 	scene_root->set_lod_threshold(lod_threshold);
+
+	RS::get_singleton()->decals_set_filter(RS::DecalFilter(int(GLOBAL_GET("rendering/textures/decals/filter"))));
```
Comment (clayjohn) [inline editor/editor_node.cpp:512]: Shouldn't we include decal filter and projector light filter here?

Comment (clayjohn) [reply editor/editor_node.cpp:512]: Nevermind. Meant to ask about the quality settings. But now I can see the quality settings is based on what sampler is used.

**Review by clayjohn — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**